### ISSUE TITLE
sum on Float Tensor and CUDA Tensor not the same

### ISSUE THREAD

**Issue by haidark** [torch/torch7#721]
I noticed this when comparing gradient parameters from identical models. The summation operations in the CPU vs GPU yield different results on the same set of numbers. Surprisingly, the following example illustrates the problem:

```
torch.setdefaulttensortype('torch.FloatTensor')
torch.manualSeed(1234)
t = torch.rand(5)
print(torch.sum(t))
c = t:cuda()
print(torch.sum(c))
```

`2.9438826441765`
`2.9438824653625`

The difference is not actually small, any explanation or fix?

**Comment by soumith** [comment]
hi, this is because of the way summations are done on the GPU.

On CPU (sort of, if 1-threaded):

out = (x1 + (x2 + (x3 + x4)))

On GPU:

out = (x1 + x2) + (x3 + x4)

These small differences in reduction operations is expected.
